types/services.d.ts

```diff
@@ -213,6 +209,206 @@ declare namespace Realm {
+                operationType: OperationType.Delete;
+                /** The namespace (database and collection) which the document got deleted from. */
+                ns: DocumentNamespace;
+                /** A document that contains the _id of the deleted document. */
+                documentKey: DocumentKey<T["_id"]>;
+            } & BaseChangeEvent;
+
+            /**
+             * Occurs when a collection is dropped from a database.
+             */
+            type DropEvent = {
+                /** The type of operation which was performed on the document. */
+                operationType: OperationType.Drop;
+                /** The namespace (database and collection) of the collection that got dropped. */
+                ns: DocumentNamespace;
+            } & BaseChangeEvent;
+
+            /**
+             * Occurs when a collection is renamed.
+             */
+            type RenameEvent = {
+                /** The type of operation which was performed on the document. */
+                operationType: OperationType.Rename;
+                /** The original namespace (database and collection) that got renamed. */
+                ns: DocumentNamespace;
+                /** The namespace (database and collection) going forward. */
+                to: DocumentNamespace;
+            } & BaseChangeEvent;
+
+            /**
+             * Occurs when a database is dropped.
+             */
+            type DropDatabaseEvent = {
+                /** The type of operation which was performed on the document. */
+                operationType: OperationType.DropDatabase;
+                /** The namespace (specifying the database name) of the database that got dropped. */
+                ns: Omit<DocumentNamespace, "coll">;
+            } & BaseChangeEvent;
+
+            /**
+             * Invalidate events close the change stream cursor.
+             */
+            type InvalidateEvent = BaseChangeEvent;
+
+            /**
+             * Represents a change event communicated via a MongoDB change stream.
+             *
+             * @see https://docs.mongodb.com/manual/reference/change-events/
+             */
+            type ChangeEvent<T extends Document> =
+                | InsertEvent<T>
+                | UpdateEvent<T>
+                | ReplaceEvent<T>
+                | DeleteEvent<T>
+                | DropEvent
+                | RenameEvent
+                | DropDatabaseEvent
+                | InvalidateEvent;
+
+            /**
+             * Options provided when initiating a watch.
+             */
+            type WatchOptions<IdType> =
+                | {}
+                | {
+                      /** Filter events. */
+                      filter: Filter;
+                  }
+                | {
+                      /** A list of document ids to watch. */
+                      ids: IdType[];
+                  };
```

Comment:
Turns out this doesn't actually work at preventing mixing `filter` and `ids` since you are always allowed to have extra fields (and `|{}` even means that any object can be passed, oops!) In order to disallow mixing `filter` and `ids` we would need to add `filter: never` and/or `ids: never` to each of the 3 alternatives. But that still leaves the type a bit of a pain to work with. Maybe it is best to just document the restriction and use a simple `{filter?: Filter, ids?: IdType[]}` definition.

Reply:
That is unexpected - I'll have to investigate further.

Reply:
Function overloads seem to have the desired semantics: [Playground Link](https://www.typescriptlang.org/play/?target=5#code/PTAEEsDtQEwOgC4GcBQAzArpAxg8B7aNACgEMAnAcyQH4AuUAb1AA97RIMBbAIwFNyoAL4BKBp14CA3Oiy4CRMlSQNmAT3Hd+g0ZsnkZKEBGjguABwA2iVJhx5CoEsxYBeAAwAaUGo-DQrkyiTKDkfAgY5NAsoADUPlLCKOjEIjLOounELgwAjMJpKep5BYbOrHneGqD5maDG2PgW4JZ8oALk+ORAA)

Reply:
Okay - it makes sense, I believe you're right that function overloading is the best way to express this.
It does however have the drawback that it seems I need to provide the doc comment three times.